### PR TITLE
Fix security

### PR DESCRIPTION
**Pull request by AnthonySendra**
**Bugfix**
* Always return `_source` attribute (also in nested profile/role)
* Return `NotFoundError` on `getUser`, `getProfile` and `getRole` when document with id doesn't exist
* Store in memory (cache) profiles and roles
* In functionnal testing, on `searchUsers`, `searchProfiles` and `searchRoles`, only check results for `world.idPrefix`. Avoid to mix with role/user/profil from real user data

**Clean**
* In `authController` use `getUser` for function `kuzzle.funnel.security.getCurrentUser` instead of reimplement the function

### CODE REVIEW

**Comment by stafyniaksacha** [comment]
+1

**Comment by ballinette** [comment]
not directly related to the changes you made, but I found something strange within formatProfileForSerialization and formatUserForSerialization functions:
if "hydrate" is true, the response will be overrided and all its additional attributes deleted.

Instead of,
  var formatUserForSerialization = function (user, hydrate) {
    if (hydrate) {
      return {
        _id: user._id,
        _source: {
          profile: formatProfileForSerialization(user.profile, true)
        }
      };
    }
    return user;
  };

we should have something like:

  var formatUserForSerialization = function (user, hydrate) {
    var response = { _id: user._id, _source: user }
    if (hydrate) {
      response._source.profile = formatProfileForSerialization(user.profile, true);
      };
    }
    delete response._source._id;
    return response;
  };

**Comment by ballinette** [comment]
:+1:

**Comment by AnthonySendra** [comment]
Close the PR for update it with feedback from issue #159

**Comment by scottinet** [comment]
About the last issue you raised, the one about the `authController`: I reimplemented the function because of the `security:getUser` hook. I don't think this is a good idea to fire this hook when performing a `auth:getCurrentUser` action.

To avoid repetition of code, some of the `getUser` code could be moved to the user repository, and both `auth:getCurrentUser` and `security:getUser` could invoke this code.

What do you think?

**Comment by AnthonySendra** [comment]
@scottinet Yep, good point, I'm on it.

**Comment by ballinette** [comment]
agree with @scottinet

File: test/api/controllers/securityController/users.test.js [hunk 1]
```diff
@@ -74,31 +75,12 @@ describe('Test: security controller - users', function () {
...
-
-          done();
-
-        })
-        .catch(error => { done(error); });
-    });
-
-    it('should respond found:false when the user is not found', done => {
-      kuzzle.funnel.security.getUser(new RequestObject({
+    it('should reject wit NotFoundError when the user is not found', () => {
```
Comment (dbengsch) [inline test/api/controllers/securityController/users.test.js]: typo on `with` ?

**Comment by scottinet** [comment]
+1

**Comment by AnthonySendra** [comment]
As discussed with @scottinet, the hook `auth:getCurrentUser` is removed because plugin can know (or in the future, we have to let the plugin know) who is the user and can do something specific when he tries to get himself.

**Pull request merged**